src/app/collective-rewards/user/hooks/useGetBuilders.ts

```diff
@@ -14,67 +25,104 @@ export type BuildersLoader = Omit<BuilderLoader, 'data'> & {
-      const proposals = buildersProposalsMap?.[address] ?? {}
+const getCombinedBuilderStatus = (builderState: BuilderStateStruct): BuilderStatusWithExcluded => {
+  // TODO: to review it
+  const [activated, kycApproved, communityApproved, , revoked, ,] = builderState
+  return revoked
+    ? BuilderStatusActive
+    : activated && kycApproved && communityApproved
+      ? BuilderStatusActive
+      : kycApproved || communityApproved
+        ? BuilderStatusInProgress
+        : EXCLUDED_BUILDER_STATUS // used to filter out builders
+}
 
-      return {
-        address,
-        status: isWhitelistedBuilder ? 'Whitelisted' : 'In progress',
-        proposals: Object.values(proposals),
-      } as BuilderInfo
-    }
-  }, [buildersProposalsMap, address, isWhitelistedBuilder])
+export const useGetBuilders = (): BuildersLoader => {
+  /*
+   * get Gauges
+   * for each Gauge
+   *    get Builder from Gauge
+   *    get Builder state
+   *    ignore the builder if paused or revoked (to be confirmed)
+   */
+  // get the gauges
+  const { data: gauges, isLoading: gaugesLoading, error: gaugesError } = useGetGaugesArray()
+  // get the builders for each gauge
+  const gaugeToBuilderCalls = gauges?.map(
+    gauge =>
+      ({
+        address: BackersManagerAddress,
+        abi: BuilderRegistryAbi,
+        functionName: 'gaugeToBuilder',
+        args: [gauge],
+      }) as const,
+  )
+  const {
+    data: buildersResult,
+    isLoading: buildersLoading,
+    error: buildersError,
+  } = useReadContracts<Address[]>({
+    contracts: gaugeToBuilderCalls,
+    query: {
+      refetchInterval: AVERAGE_BLOCKTIME,
+    },
+  })
+  const builders = buildersResult?.map(builder => builder.result) as Address[]
 
-  const isLoading = builderProposalsMapLoading || isWhitelistedBuilderLoading
-  const error = builderProposalsMapError ?? isWhitelistedBuilderError
+  // get the builder state for each builder
+  const builderStatesCalls = builders?.map(
+    builder =>
+      ({
+        address: BackersManagerAddress,
+        abi: BuilderRegistryAbi,
+        functionName: 'builderState',
+        args: [builder],
+      }) as const,
+  )
+  const {
+    data: builderStatesResult,
+    isLoading: builderStatesLoading,
+    error: builderStatesError,
+  } = useReadContracts({ contracts: builderStatesCalls, query: { refetchInterval: AVERAGE_BLOCKTIME } })
+  const builderStates = builderStatesResult?.map(builderState => builderState.result as BuilderStateStruct)
 
-  return {
-    data,
-    isLoading,
-    error,
-  }
-}
+  const builderStatusMap = builders
+    ?.map((builder, index) => {
+      return {
+        address: builder,
+        status: getCombinedBuilderStatus((builderStates?.[index] ?? []) as BuilderStateStruct),
+      }
+    })
+    .filter(builder => builder.status !== EXCLUDED_BUILDER_STATUS)
+    .reduce<BuilderStatusMap>((acc, builder) => {
+      acc[builder.address] = builder.status as BuilderStatus
+      return acc
+    }, {})
```

Comment:
why do it in three if you're reducing at the end anyway? I'd just reduce

Reply:
Although not required, I find it more readable splitting it in 3 steps.

Reply:
really?

```ts
const builderStatusMap = builders?.reduce<BuilderStatusMap>((acc, builder, index) => {
    const builderState = (builderStates?.[index] ?? []) as BuilderStateStruct
    const status = getCombinedBuilderStatus(builderState)
    if (status !== EXCLUDED_BUILDER_STATUS) {
      acc[builder] = status as BuilderStatus
    }

    return acc
  }, {});
```

Reply:
the builders array may be quite large and while O(3n) is just O(n), in practice it could be taxing to iterate 3 times over it (well twice over a subset granted, but still)